In [2]:
!unzip /content/build.zip

Archive:  /content/build.zip
   creating: content/llama.cpp/build/
   creating: content/llama.cpp/build/Testing/
   creating: content/llama.cpp/build/Testing/Temporary/
  inflating: content/llama.cpp/build/DartConfiguration.tcl  
  inflating: content/llama.cpp/build/CMakeCache.txt  
  inflating: content/llama.cpp/build/CTestTestfile.cmake  
   creating: content/llama.cpp/build/CMakeFiles/
   creating: content/llama.cpp/build/CMakeFiles/FindOpenMP/
  inflating: content/llama.cpp/build/CMakeFiles/FindOpenMP/ompver_C.bin  
  inflating: content/llama.cpp/build/CMakeFiles/FindOpenMP/ompver_CXX.bin  
   creating: content/llama.cpp/build/CMakeFiles/ExperimentalStart.dir/
  inflating: content/llama.cpp/build/CMakeFiles/ExperimentalStart.dir/compiler_depend.ts  
  inflating: content/llama.cpp/build/CMakeFiles/ExperimentalStart.dir/compiler_depend.make  
 extracting: content/llama.cpp/build/CMakeFiles/ExperimentalStart.dir/progress.make  
  inflating: content/llama.cpp/build/CMakeFiles/Experimen

In [4]:
!/content/llama.cpp/build/bin/llama-cli -h

----- common params -----

-h,    --help, --usage                  print usage and exit
--version                               show version and build info
--completion-bash                       print source-able bash completion script for llama.cpp
--verbose-prompt                        print a verbose prompt before generation (default: false)
-t,    --threads N                      number of threads to use during generation (default: -1)
                                        (env: LLAMA_ARG_THREADS)
-tb,   --threads-batch N                number of threads to use during batch and prompt processing (default:
                                        same as --threads)
-C,    --cpu-mask M                     CPU affinity mask: arbitrarily long hex. Complements cpu-range
                                        (default: "")
-Cr,   --cpu-range lo-hi                range of CPUs for affinity. Complements --cpu-mask
--cpu-strict <0|1>                      use strict CPU placement (defaul

In [5]:
!wget https://huggingface.co/bartowski/Llama-3.2-3B-Instruct-GGUF/resolve/main/Llama-3.2-3B-Instruct-Q4_K_M.gguf

--2025-04-16 02:52:51--  https://huggingface.co/bartowski/Llama-3.2-3B-Instruct-GGUF/resolve/main/Llama-3.2-3B-Instruct-Q4_K_M.gguf
Resolving huggingface.co (huggingface.co)... 3.165.160.59, 3.165.160.12, 3.165.160.61, ...
Connecting to huggingface.co (huggingface.co)|3.165.160.59|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://cas-bridge.xethub.hf.co/xet-bridge-us/66f457f561dbc064f4a6413f/4cb1444f81355e47d236ada8190f0325ce46412a83f3ab62a1d63bb314592ebc?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=cas%2F20250416%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250416T025251Z&X-Amz-Expires=3600&X-Amz-Signature=6f0774e2af72fefa5e246ae2d6bec453493b747ce94f4a3aa6172342e321973d&X-Amz-SignedHeaders=host&X-Xet-Cas-Uid=public&response-content-disposition=inline%3B+filename*%3DUTF-8%27%27Llama-3.2-3B-Instruct-Q4_K_M.gguf%3B+filename%3D%22Llama-3.2-3B-Instruct-Q4_K_M.gguf%22%3B&x-id=GetObject&Expires=1744775571&Policy=

In [7]:
!/content/llama.cpp/build/bin/llama-cli -m /content/Llama-3.2-3B-Instruct-Q4_K_M.gguf -p "What is a car?"

build: 5142 (80f19b41) with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu
main: llama backend init
main: load the model and apply lora adapter, if any
llama_model_loader: loaded meta data with 35 key-value pairs and 255 tensors from /content/Llama-3.2-3B-Instruct-Q4_K_M.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.type str              = model
llama_model_loader: - kv   2:                               general.name str              = Llama 3.2 3B Instruct
llama_model_loader: - kv   3:                           general.finetune str              = Instruct
llama_model_loader: - kv   4:                           general.basename str              = Llama-3.2
llama_model_loader: - kv   5:                         general.size_lab

In [8]:
!/content/llama.cpp/build/bin/llama-server -m /content/Llama-3.2-3B-Instruct-Q4_K_M.gguf

build: 5142 (80f19b41) with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu
system info: n_threads = 1, n_threads_batch = 1, total_threads = 2

system_info: n_threads = 1 (n_threads_batch = 1) / 2 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | LLAMAFILE = 1 | OPENMP = 1 | AARCH64_REPACK = 1 | 

main: binding port with default address family
main: couldn't bind HTTP server socket, hostname: 127.0.0.1, port: 8080
srv    operator(): operator(): cleaning up before exit...


In [9]:
!pip install pyngrok

In [1]:
from pyngrok import ngrok
import threading
import time

ngrok_token = "2qoMyvN33sChLDfKL39OR8IDIxk_2hoKPdPQp4MUdSQbDPv6d"
ngrok.set_auth_token(ngrok_token)

# Start ngrok in a separate thread to avoid blocking
def start_ngrok():
    public_url = ngrok.connect(8080).public_url
    print(f"🚀 Ngrok Tunnel Open: {public_url}")

ngrok_thread = threading.Thread(target=start_ngrok)
ngrok_thread.start()

# Wait for ngrok to start (optional)
time.sleep(5)

# Execute your node.js script
!/content/llama.cpp/build/bin/llama-server -m /content/Llama-3.2-3B-Instruct-Q4_K_M.gguf

🚀 Ngrok Tunnel Open: https://fba3-35-197-34-232.ngrok-free.app
build: 5142 (80f19b41) with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu
system info: n_threads = 1, n_threads_batch = 1, total_threads = 2

system_info: n_threads = 1 (n_threads_batch = 1) / 2 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | LLAMAFILE = 1 | OPENMP = 1 | AARCH64_REPACK = 1 | 

main: binding port with default address family
main: couldn't bind HTTP server socket, hostname: 127.0.0.1, port: 8080
srv    operator(): operator(): cleaning up before exit...


In [ ]:
!pip install pyngrok

2qoMyvN33sChLDfKL39OR8IDIxk_2hoKPdPQp4MUdSQbDPv6d


from pyngrok import ngrok
import threading
import time

ngrok_token = "2qoMyvN33sChLDfKL39OR8IDIxk_2hoKPdPQp4MUdSQbDPv6d"
ngrok.set_auth_token(ngrok_token)

# Start ngrok in a separate thread to avoid blocking
def start_ngrok():
    public_url = ngrok.connect(3000).public_url
    print(f"🚀 Ngrok Tunnel Open: {public_url}")

ngrok_thread = threading.Thread(target=start_ngrok)
ngrok_thread.start()

# Wait for ngrok to start (optional)
time.sleep(5)

# Execute your node.js script
!node hello-world.js

In [4]:
!/content/llama.cpp/build/bin/llama-server -m /content/Llama-3.2-3B-Instruct-Q4_K_M.gguf

build: 5142 (80f19b41) with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu
system info: n_threads = 1, n_threads_batch = 1, total_threads = 2

system_info: n_threads = 1 (n_threads_batch = 1) / 2 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | LLAMAFILE = 1 | OPENMP = 1 | AARCH64_REPACK = 1 | 

main: binding port with default address family
main: couldn't bind HTTP server socket, hostname: 127.0.0.1, port: 8080
srv    operator(): operator(): cleaning up before exit...


In [5]:
public_url = ngrok.connect(8080).public_url  # Replace 3000 with your chosen port
print(f"🚀 Ngrok Tunnel Open: {public_url}")

🚀 Ngrok Tunnel Open: https://a051-35-197-34-232.ngrok-free.app


In [1]:
# -*- coding: utf-8 -*-
import os
from pyngrok import ngrok, conf
import threading
import time

# ----- إعدادات -----
NGROK_TOKEN = "2qoMyvN33sChLDfKL39OR8IDIxk_2hoKPdPQp4MUdSQbDPv6d" # استبدله بالتوكن الخاص بك أو استخدم Colab Secrets
MODEL_PATH = "/content/Llama-3.2-3B-Instruct-Q4_K_M.gguf"        # مسار الموديل الذي تم تحميله
LLAMA_SERVER_PATH = "/content/llama.cpp/build/bin/llama-server"  # مسار الملف التنفيذي للخادم
PORT = 8080                                                     # المنفذ الذي سيعمل عليه الخادم (الافتراضي لـ llama-server)
HOST = "0.0.0.0"                                                # للاستماع على جميع الواجهات

# ----- إعداد Ngrok -----
# يفضل استخدام Colab Secrets:
# from google.colab import userdata
# NGROK_TOKEN = userdata.get('NGROK_AUTH_TOKEN') # اسم السيكرت يجب أن يكون NGROK_AUTH_TOKEN

if not NGROK_TOKEN:
    print("⚠️ يرجى تعيين توكن Ngrok الخاص بك في متغير NGROK_TOKEN أو عبر Colab Secrets باسم 'NGROK_AUTH_TOKEN'")
else:
    try:
        conf.get_default().auth_token = NGROK_TOKEN
        print("🔑 تم تعيين توكن المصادقة لـ Ngrok.")
    except Exception as e:
        print(f"خطأ في إعداد توكن Ngrok: {e}")
        # يمكنك إيقاف التنفيذ هنا إذا كان التوكن ضرورياً
        # raise SystemExit()

# متغير لتخزين رابط Ngrok العام
public_url = None
ngrok_thread_started = threading.Event() # للإشارة عند بدء تشغيل نفق ngrok

# وظيفة لبدء Ngrok في thread منفصل
def start_ngrok(port):
    global public_url
    try:
        # قطع أي أنفاق قديمة قد تكون نشطة لنفس التوكن
        for tunnel in ngrok.get_tunnels():
            ngrok.disconnect(tunnel.public_url)
            print(f"🔌 تم قطع نفق Ngrok قديم: {tunnel.public_url}")
        # بدء نفق جديد
        tunnel = ngrok.connect(port, "http")
        public_url = tunnel.public_url
        print(f"🚀 نفق Ngrok يعمل: {public_url}")
        print(f"🔗 يمكن الوصول للخادم عبر هذا الرابط.")
        ngrok_thread_started.set() # إرسال إشارة بأن النفق جاهز
        # إبقاء الـ thread يعمل حتى يتم إيقاف البرنامج الرئيسي
        while True:
            time.sleep(1)
    except Exception as e:
        print(f"❌ خطأ أثناء تشغيل Ngrok: {e}")
        public_url = f"Error: {e}"
        ngrok_thread_started.set() # إرسال إشارة حتى لو حدث خطأ لتجنب الانتظار للأبد

# ----- تشغيل Ngrok -----
if NGROK_TOKEN: # فقط إذا كان التوكن موجودًا
    print("⏳ بدء تشغيل Ngrok في الخلفية...")
    ngrok_thread = threading.Thread(target=start_ngrok, args=(PORT,))
    ngrok_thread.daemon = True # اجعل الـ thread يموت عند انتهاء البرنامج الرئيسي
    ngrok_thread.start()

    # الانتظار حتى يبدأ Ngrok ويحصل على الرابط أو يحدث خطأ
    print("⏳ الانتظار لبدء نفق Ngrok...")
    started = ngrok_thread_started.wait(timeout=15) # انتظر 15 ثانية كحد أقصى

    if not started:
        print("⚠️ استغرق Ngrok وقتًا طويلاً للبدء أو فشل. قد لا يعمل الخادم بشكل صحيح عبر الإنترنت.")
    elif public_url and "Error" in public_url:
         print(f"❌ فشل Ngrok في البدء بشكل صحيح: {public_url}")
         # يمكنك اختيار إيقاف الخلية هنا إذا أردت
         # raise SystemExit("فشل Ngrok")
    else:
        print("✅ Ngrok جاهز.")

# ----- تشغيل خادم Llama -----
# تأكد من أن الملفات موجودة قبل التشغيل
if not os.path.exists(LLAMA_SERVER_PATH):
    print(f"❌ خطأ: لم يتم العثور على الملف التنفيذي للخادم في {LLAMA_SERVER_PATH}")
    print("❓ هل قمت بتشغيل خطوات البناء في الخطوة 1؟")
elif not os.path.exists(MODEL_PATH):
    print(f"❌ خطأ: لم يتم العثور على ملف الموديل في {MODEL_PATH}")
    print("❓ هل قمت بتحميل الموديل إلى المسار الصحيح؟")
else:
    print(f"\n🚀 بدء تشغيل llama-server على المنفذ {PORT}...")
    print(f"   باستخدام الموديل: {MODEL_PATH}")

    # بناء أمر التشغيل
    # أضف وسائط أخرى حسب الحاجة، مثل:
    # -c 2048 (حجم السياق)
    # -ngl 35 (عدد الطبقات لتشغيلها على GPU - إذا تم بناء llama.cpp مع دعم GPU)
    # --host 0.0.0.0 (للاستماع على جميع الواجهات)
    # --port 8080 (لتحديد المنفذ)
    command = f"{LLAMA_SERVER_PATH} -m {MODEL_PATH} --host {HOST} --port {PORT} -c 2048" # مثال مع حجم سياق

    # أضف خيار GPU إذا كان متاحًا (اضبط العدد حسب ذاكرة الـ VRAM لديك)
    # command += " -ngl 35" # مثال لـ 35 طبقة على GPU

    print(f"\n✨ الأمر الذي سيتم تنفيذه:\n   {command}\n")

    # تنفيذ الأمر (سيوقف هذا الخلية حتى توقف الخادم يدوياً)
    !{command}

    # الكود هنا لن يتم تنفيذه إلا بعد إيقاف الخادم (مثل الضغط على زر الإيقاف في Colab)
    print("\n🛑 تم إيقاف خادم Llama.")

    # إيقاف Ngrok عند إيقاف الخادم (اختياري، الـ thread سينتهي تلقائياً إذا كان daemon)
    try:
        ngrok.disconnect(public_url)
        print("🔌 تم قطع اتصال نفق Ngrok.")
    except Exception as e:
        print(f"⚠️ لم يتمكن من قطع اتصال Ngrok أو أنه قُطع بالفعل: {e}")

🔑 تم تعيين توكن المصادقة لـ Ngrok.
⏳ بدء تشغيل Ngrok في الخلفية...
⏳ الانتظار لبدء نفق Ngrok...
🚀 نفق Ngrok يعمل: https://0d1d-35-197-34-232.ngrok-free.app
🔗 يمكن الوصول للخادم عبر هذا الرابط.
✅ Ngrok جاهز.

🚀 بدء تشغيل llama-server على المنفذ 8080...
   باستخدام الموديل: /content/Llama-3.2-3B-Instruct-Q4_K_M.gguf

✨ الأمر الذي سيتم تنفيذه:
   /content/llama.cpp/build/bin/llama-server -m /content/Llama-3.2-3B-Instruct-Q4_K_M.gguf --host 0.0.0.0 --port 8080 -c 2048

build: 5142 (80f19b41) with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu
system info: n_threads = 1, n_threads_batch = 1, total_threads = 2

system_info: n_threads = 1 (n_threads_batch = 1) / 2 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | LLAMAFILE = 1 | OPENMP = 1 | AARCH64_REPACK = 1 | 

main: binding port with default address family
main: couldn't bind HTTP server socket, hostname: 0.0.0.0, port: 8080
srv    operator(): operator(): cleaning up before exit...



🛑 تم إيقاف خادم Llama.
🔌 تم قطع اتصال نفق Ngrok.


شغال

In [2]:
# -*- coding: utf-8 -*-
import os
from pyngrok import ngrok, conf
import threading
import time

# ----- إعدادات -----
NGROK_TOKEN = "2qoMyvN33sChLDfKL39OR8IDIxk_2hoKPdPQp4MUdSQbDPv6d"
MODEL_PATH = "/content/Llama-3.2-3B-Instruct-Q4_K_M.gguf"
LLAMA_SERVER_PATH = "/content/llama.cpp/build/bin/llama-server"
PORT = 8081  # <---- تم تغيير المنفذ هنا
HOST = "0.0.0.0"

# ----- إعداد Ngrok -----
if not NGROK_TOKEN:
    print("⚠️ يرجى تعيين توكن Ngrok الخاص بك...")
else:
    try:
        conf.get_default().auth_token = NGROK_TOKEN
        print("🔑 تم تعيين توكن المصادقة لـ Ngrok.")
    except Exception as e:
        print(f"خطأ في إعداد توكن Ngrok: {e}")

public_url = None
ngrok_thread_started = threading.Event()

def start_ngrok(port_to_forward): # تغيير اسم المتغير لتوضيح أنه المنفذ المراد توجيهه
    global public_url
    try:
        for tunnel in ngrok.get_tunnels():
            ngrok.disconnect(tunnel.public_url)
            print(f"🔌 تم قطع نفق Ngrok قديم: {tunnel.public_url}")
        tunnel = ngrok.connect(port_to_forward, "http") # <---- استخدام المنفذ الجديد هنا
        public_url = tunnel.public_url
        print(f"🚀 نفق Ngrok يعمل على المنفذ {port_to_forward}: {public_url}")
        print(f"🔗 يمكن الوصول للخادم عبر هذا الرابط.")
        ngrok_thread_started.set()
        while True:
            time.sleep(1)
    except Exception as e:
        print(f"❌ خطأ أثناء تشغيل Ngrok: {e}")
        public_url = f"Error: {e}"
        ngrok_thread_started.set()

# ----- تشغيل Ngrok -----
if NGROK_TOKEN:
    print("⏳ بدء تشغيل Ngrok في الخلفية...")
    ngrok_thread = threading.Thread(target=start_ngrok, args=(PORT,)) # تمرير المنفذ الجديد
    ngrok_thread.daemon = True
    ngrok_thread.start()
    print("⏳ الانتظار لبدء نفق Ngrok...")
    started = ngrok_thread_started.wait(timeout=15)
    if not started:
        print("⚠️ استغرق Ngrok وقتًا طويلاً للبدء أو فشل.")
    elif public_url and "Error" in public_url:
         print(f"❌ فشل Ngrok في البدء بشكل صحيح: {public_url}")
    else:
        print("✅ Ngrok جاهز.")

# ----- تشغيل خادم Llama -----
if not os.path.exists(LLAMA_SERVER_PATH):
    print(f"❌ خطأ: لم يتم العثور على الملف التنفيذي للخادم في {LLAMA_SERVER_PATH}")
elif not os.path.exists(MODEL_PATH):
    print(f"❌ خطأ: لم يتم العثور على ملف الموديل في {MODEL_PATH}")
else:
    print(f"\n🚀 بدء تشغيل llama-server على المنفذ {PORT}...") # طباعة المنفذ الجديد
    print(f"   باستخدام الموديل: {MODEL_PATH}")

    command = f"{LLAMA_SERVER_PATH} -m {MODEL_PATH} --host {HOST} --port {PORT} -c 2048" # <---- استخدام المنفذ الجديد هنا

    print(f"\n✨ الأمر الذي سيتم تنفيذه:\n   {command}\n")
    !{command}
    print("\n🛑 تم إيقاف خادم Llama.")
    try:
        if public_url and not "Error" in public_url:
             ngrok.disconnect(public_url)
             print("🔌 تم قطع اتصال نفق Ngrok.")
    except Exception as e:
        print(f"⚠️ لم يتمكن من قطع اتصال Ngrok أو أنه قُطع بالفعل: {e}")

🔑 تم تعيين توكن المصادقة لـ Ngrok.
⏳ بدء تشغيل Ngrok في الخلفية...
⏳ الانتظار لبدء نفق Ngrok...
🚀 نفق Ngrok يعمل على المنفذ 8081: https://b33a-35-197-34-232.ngrok-free.app
🔗 يمكن الوصول للخادم عبر هذا الرابط.
✅ Ngrok جاهز.

🚀 بدء تشغيل llama-server على المنفذ 8081...
   باستخدام الموديل: /content/Llama-3.2-3B-Instruct-Q4_K_M.gguf

✨ الأمر الذي سيتم تنفيذه:
   /content/llama.cpp/build/bin/llama-server -m /content/Llama-3.2-3B-Instruct-Q4_K_M.gguf --host 0.0.0.0 --port 8081 -c 2048

build: 5142 (80f19b41) with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu
system info: n_threads = 1, n_threads_batch = 1, total_threads = 2

system_info: n_threads = 1 (n_threads_batch = 1) / 2 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | LLAMAFILE = 1 | OPENMP = 1 | AARCH64_REPACK = 1 | 

main: binding port with default address family
main: HTTP server is listening, hostname: 0.0.0.0, port: 8081, http threads: 3
main: loading model
srv    load_model: lo

srv    operator(): operator(): cleaning up before exit...
Received second interrupt, terminating immediately.

🛑 تم إيقاف خادم Llama.
🔌 تم قطع اتصال نفق Ngrok.


🔑 تم تعيين توكن المصادقة لـ Ngrok.
⏳ بدء تشغيل Ngrok في الخلفية...
⏳ الانتظار لبدء نفق Ngrok...
🚀 نفق Ngrok يعمل على المنفذ 8081: https://b33a-35-197-34-232.ngrok-free.app
🔗 يمكن الوصول للخادم عبر هذا الرابط.
✅ Ngrok جاهز.

🚀 بدء تشغيل llama-server على المنفذ 8081...
   باستخدام الموديل: /content/Llama-3.2-3B-Instruct-Q4_K_M.gguf

✨ الأمر الذي سيتم تنفيذه:
   /content/llama.cpp/build/bin/llama-server -m /content/Llama-3.2-3B-Instruct-Q4_K_M.gguf --host 0.0.0.0 --port 8081 -c 2048

build: 5142 (80f19b41) with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu
system info: n_threads = 1, n_threads_batch = 1, total_threads = 2

system_info: n_threads = 1 (n_threads_batch = 1) / 2 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | LLAMAFILE = 1 | OPENMP = 1 | AARCH64_REPACK = 1 |

main: binding port with default address family
main: HTTP server is listening, hostname: 0.0.0.0, port: 8081, http threads: 3
main: loading model
srv    load_model: loading model '/content/Llama-3.2-3B-Instruct-Q4_K_M.gguf'
llama_model_loader: loaded meta data with 35 key-value pairs and 255 tensors from /content/Llama-3.2-3B-Instruct-Q4_K_M.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.type str              = model
llama_model_loader: - kv   2:                               general.name str              = Llama 3.2 3B Instruct
llama_model_loader: - kv   3:                           general.finetune str              = Instruct
llama_model_loader: - kv   4:                           general.basename str              = Llama-3.2
llama_model_loader: - kv   5:                         general.size_label str              = 3B
llama_model_loader: - kv   6:                            general.license str              = llama3.2
llama_model_loader: - kv   7:                               general.tags arr[str,6]       = ["facebook", "meta", "pytorch", "llam...
llama_model_loader: - kv   8:                          general.languages arr[str,8]       = ["en", "de", "fr", "it", "pt", "hi", ...
llama_model_loader: - kv   9:                          llama.block_count u32              = 28
llama_model_loader: - kv  10:                       llama.context_length u32              = 131072
llama_model_loader: - kv  11:                     llama.embedding_length u32              = 3072
llama_model_loader: - kv  12:                  llama.feed_forward_length u32              = 8192
llama_model_loader: - kv  13:                 llama.attention.head_count u32              = 24
llama_model_loader: - kv  14:              llama.attention.head_count_kv u32              = 8
llama_model_loader: - kv  15:                       llama.rope.freq_base f32              = 500000.000000
llama_model_loader: - kv  16:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  17:                 llama.attention.key_length u32              = 128
llama_model_loader: - kv  18:               llama.attention.value_length u32              = 128
llama_model_loader: - kv  19:                          general.file_type u32              = 15
llama_model_loader: - kv  20:                           llama.vocab_size u32              = 128256
llama_model_loader: - kv  21:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv  22:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  23:                         tokenizer.ggml.pre str              = llama-bpe
llama_model_loader: - kv  24:                      tokenizer.ggml.tokens arr[str,128256]  = ["!", "\"", "#", "$", "%", "&", "'", ...
llama_model_loader: - kv  25:                  tokenizer.ggml.token_type arr[i32,128256]  = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...
llama_model_loader: - kv  26:                      tokenizer.ggml.merges arr[str,280147]  = ["Ġ Ġ", "Ġ ĠĠĠ", "ĠĠ ĠĠ", "...
llama_model_loader: - kv  27:                tokenizer.ggml.bos_token_id u32              = 128000
llama_model_loader: - kv  28:                tokenizer.ggml.eos_token_id u32              = 128009
llama_model_loader: - kv  29:                    tokenizer.chat_template str              = {{- bos_token }}\n{%- if custom_tools ...
llama_model_loader: - kv  30:               general.quantization_version u32              = 2
llama_model_loader: - kv  31:                      quantize.imatrix.file str              = /models_out/Llama-3.2-3B-Instruct-GGU...
llama_model_loader: - kv  32:                   quantize.imatrix.dataset str              = /training_dir/calibration_datav3.txt
llama_model_loader: - kv  33:             quantize.imatrix.entries_count i32              = 196
llama_model_loader: - kv  34:              quantize.imatrix.chunks_count i32              = 125
llama_model_loader: - type  f32:   58 tensors
llama_model_loader: - type q4_K:  168 tensors
llama_model_loader: - type q6_K:   29 tensors
print_info: file format = GGUF V3 (latest)
print_info: file type   = Q4_K - Medium
print_info: file size   = 1.87 GiB (5.01 BPW)
load: special tokens cache size = 256
load: token to piece cache size = 0.7999 MB
print_info: arch             = llama
print_info: vocab_only       = 0
print_info: n_ctx_train      = 131072
print_info: n_embd           = 3072
print_info: n_layer          = 28
print_info: n_head           = 24
print_info: n_head_kv        = 8
print_info: n_rot            = 128
print_info: n_swa            = 0
print_info: n_swa_pattern    = 1
print_info: n_embd_head_k    = 128
print_info: n_embd_head_v    = 128
print_info: n_gqa            = 3
print_info: n_embd_k_gqa     = 1024
print_info: n_embd_v_gqa     = 1024
print_info: f_norm_eps       = 0.0e+00
print_info: f_norm_rms_eps   = 1.0e-05
print_info: f_clamp_kqv      = 0.0e+00
print_info: f_max_alibi_bias = 0.0e+00
print_info: f_logit_scale    = 0.0e+00
print_info: f_attn_scale     = 0.0e+00
print_info: n_ff             = 8192
print_info: n_expert         = 0
print_info: n_expert_used    = 0
print_info: causal attn      = 1
print_info: pooling type     = 0
print_info: rope type        = 0
print_info: rope scaling     = linear
print_info: freq_base_train  = 500000.0
print_info: freq_scale_train = 1
print_info: n_ctx_orig_yarn  = 131072
print_info: rope_finetuned   = unknown
print_info: ssm_d_conv       = 0
print_info: ssm_d_inner      = 0
print_info: ssm_d_state      = 0
print_info: ssm_dt_rank      = 0
print_info: ssm_dt_b_c_rms   = 0
print_info: model type       = 3B
print_info: model params     = 3.21 B
print_info: general.name     = Llama 3.2 3B Instruct
print_info: vocab type       = BPE
print_info: n_vocab          = 128256
print_info: n_merges         = 280147
print_info: BOS token        = 128000 '<|begin_of_text|>'
print_info: EOS token        = 128009 '<|eot_id|>'
print_info: EOT token        = 128009 '<|eot_id|>'
print_info: EOM token        = 128008 '<|eom_id|>'
print_info: LF token         = 198 'Ċ'
print_info: EOG token        = 128008 '<|eom_id|>'
print_info: EOG token        = 128009 '<|eot_id|>'
print_info: max token length = 256
load_tensors: loading model tensors, this can take a while... (mmap = true)
load_tensors:  CPU_AARCH64 model buffer size =  1299.38 MiB
load_tensors:   CPU_Mapped model buffer size =  1918.35 MiB
....................................................................................
llama_context: constructing llama_context
llama_context: n_seq_max     = 1
llama_context: n_ctx         = 2048
llama_context: n_ctx_per_seq = 2048
llama_context: n_batch       = 2048
llama_context: n_ubatch      = 512
llama_context: causal_attn   = 1
llama_context: flash_attn    = 0
llama_context: freq_base     = 500000.0
llama_context: freq_scale    = 1
llama_context: n_ctx_per_seq (2048) < n_ctx_train (131072) -- the full capacity of the model will not be utilized
llama_context:        CPU  output buffer size =     0.49 MiB
init: kv_size = 2048, offload = 1, type_k = 'f16', type_v = 'f16', n_layer = 28, can_shift = 1
init:        CPU KV buffer size =   224.00 MiB
llama_context: KV self size  =  224.00 MiB, K (f16):  112.00 MiB, V (f16):  112.00 MiB
llama_context:        CPU compute buffer size =   256.50 MiB
llama_context: graph nodes  = 958
llama_context: graph splits = 1
common_init_from_params: setting dry_penalty_last_n to ctx_size = 2048
common_init_from_params: warming up the model with an empty run - please wait ... (--no-warmup to disable)
srv          init: initializing slots, n_slots = 1
slot         init: id  0 | task -1 | new slot n_ctx_slot = 2048
main: model loaded
main: chat template, chat_template: {{- bos_token }}
{%- if custom_tools is defined %}
    {%- set tools = custom_tools %}
{%- endif %}
{%- if not tools_in_user_message is defined %}
    {%- set tools_in_user_message = true %}
{%- endif %}
{%- if not date_string is defined %}
    {%- if strftime_now is defined %}
        {%- set date_string = strftime_now("%d %b %Y") %}
    {%- else %}
        {%- set date_string = "26 Jul 2024" %}
    {%- endif %}
{%- endif %}
{%- if not tools is defined %}
    {%- set tools = none %}
{%- endif %}

{#- This block extracts the system message, so we can slot it into the right place. #}
{%- if messages[0]['role'] == 'system' %}
    {%- set system_message = messages[0]['content']|trim %}
    {%- set messages = messages[1:] %}
{%- else %}
    {%- set system_message = "" %}
{%- endif %}

{#- System message #}
{{- "<|start_header_id|>system<|end_header_id|>\n\n" }}
{%- if tools is not none %}
    {{- "Environment: ipython\n" }}
{%- endif %}
{{- "Cutting Knowledge Date: December 2023\n" }}
{{- "Today Date: " + date_string + "\n\n" }}
{%- if tools is not none and not tools_in_user_message %}
    {{- "You have access to the following functions. To call a function, please respond with JSON for a function call." }}
    {{- 'Respond in the format {"name": function name, "parameters": dictionary of argument name and its value}.' }}
    {{- "Do not use variables.\n\n" }}
    {%- for t in tools %}
        {{- t | tojson(indent=4) }}
        {{- "\n\n" }}
    {%- endfor %}
{%- endif %}
{{- system_message }}
{{- "<|eot_id|>" }}

{#- Custom tools are passed in a user message with some extra guidance #}
{%- if tools_in_user_message and not tools is none %}
    {#- Extract the first user message so we can plug it in here #}
    {%- if messages | length != 0 %}
        {%- set first_user_message = messages[0]['content']|trim %}
        {%- set messages = messages[1:] %}
    {%- else %}
        {{- raise_exception("Cannot put tools in the first user message when there's no first user message!") }}
{%- endif %}
    {{- '<|start_header_id|>user<|end_header_id|>\n\n' -}}
    {{- "Given the following functions, please respond with a JSON for a function call " }}
    {{- "with its proper arguments that best answers the given prompt.\n\n" }}
    {{- 'Respond in the format {"name": function name, "parameters": dictionary of argument name and its value}.' }}
    {{- "Do not use variables.\n\n" }}
    {%- for t in tools %}
        {{- t | tojson(indent=4) }}
        {{- "\n\n" }}
    {%- endfor %}
    {{- first_user_message + "<|eot_id|>"}}
{%- endif %}

{%- for message in messages %}
    {%- if not (message.role == 'ipython' or message.role == 'tool' or 'tool_calls' in message) %}
        {{- '<|start_header_id|>' + message['role'] + '<|end_header_id|>\n\n'+ message['content'] | trim + '<|eot_id|>' }}
    {%- elif 'tool_calls' in message %}
        {%- if not message.tool_calls|length == 1 %}
            {{- raise_exception("This model only supports single tool-calls at once!") }}
        {%- endif %}
        {%- set tool_call = message.tool_calls[0].function %}
        {{- '<|start_header_id|>assistant<|end_header_id|>\n\n' -}}
        {{- '{"name": "' + tool_call.name + '", ' }}
        {{- '"parameters": ' }}
        {{- tool_call.arguments | tojson }}
        {{- "}" }}
        {{- "<|eot_id|>" }}
    {%- elif message.role == "tool" or message.role == "ipython" %}
        {{- "<|start_header_id|>ipython<|end_header_id|>\n\n" }}
        {%- if message.content is mapping or message.content is iterable %}
            {{- message.content | tojson }}
        {%- else %}
            {{- message.content }}
        {%- endif %}
        {{- "<|eot_id|>" }}
    {%- endif %}
{%- endfor %}
{%- if add_generation_prompt %}
    {{- '<|start_header_id|>assistant<|end_header_id|>\n\n' }}
{%- endif %}
, example_format: '<|start_header_id|>system<|end_header_id|>

You are a helpful assistant<|eot_id|><|start_header_id|>user<|end_header_id|>

Hello<|eot_id|><|start_header_id|>assistant<|end_header_id|>

Hi there<|eot_id|><|start_header_id|>user<|end_header_id|>

How are you?<|eot_id|><|start_header_id|>assistant<|end_header_id|>

'
main: server is listening on http://0.0.0.0:8081 - starting the main loop
srv  update_slots: all slots are idle
srv  log_server_r: request: GET / 127.0.0.1 200
srv  log_server_r: request: GET /favicon.ico 127.0.0.1 404
srv  params_from_: Chat format: Content-only
slot launch_slot_: id  0 | task 0 | processing task
slot update_slots: id  0 | task 0 | new prompt, n_ctx_slot = 2048, n_keep = 0, n_prompt_tokens = 22
slot update_slots: id  0 | task 0 | kv cache rm [0, end)
slot update_slots: id  0 | task 0 | prompt processing progress, n_past = 22, n_tokens = 22, progress = 1.000000
slot update_slots: id  0 | task 0 | prompt done, n_past = 22, n_tokens = 22
slot      release: id  0 | task 0 | stop processing: n_past = 31, truncated = 0
slot print_timing: id  0 | task 0 |
prompt eval time =    3500.99 ms /    22 tokens (  159.14 ms per token,     6.28 tokens per second)
       eval time =    2978.29 ms /    10 tokens (  297.83 ms per token,     3.36 tokens per second)
      total time =    6479.29 ms /    32 tokens
srv  update_slots: all slots are idle
srv  log_server_r: request: POST /v1/chat/completions 127.0.0.1 200
srv  params_from_: Chat format: Content-only
slot launch_slot_: id  0 | task 11 | processing task
slot update_slots: id  0 | task 11 | new prompt, n_ctx_slot = 2048, n_keep = 0, n_prompt_tokens = 48
slot update_slots: id  0 | task 11 | kv cache rm [31, end)
slot update_slots: id  0 | task 11 | prompt processing progress, n_past = 48, n_tokens = 17, progress = 0.354167
slot update_slots: id  0 | task 11 | prompt done, n_past = 48, n_tokens = 17

In [3]:
!lsof -i :8080

COMMAND PID USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
node      7 root   21u  IPv6  19224      0t0  TCP *:8080 (LISTEN)
node      7 root   26u  IPv6  25174      0t0  TCP 6d0a0afb4e88:8080->172.28.0.1:39814 (ESTABLISHED)
node      7 root   28u  IPv6 123113      0t0  TCP 6d0a0afb4e88:8080->172.28.0.1:53682 (ESTABLISHED)
node      7 root   30u  IPv6 122780      0t0  TCP 6d0a0afb4e88:8080->172.28.0.1:53688 (ESTABLISHED)
